client/driver/executor/executor_linux.go

```diff
@@ -221,181 +365,240 @@ func (e *UniversalExecutor) Stats() (*cstructs.TaskResourceUsage, error) {
 }
 
-// runAs takes a user id as a string and looks up the user, and sets the command
-// to execute as that user.
-func (e *UniversalExecutor) runAs(userid string) error {
-	u, err := user.Lookup(userid)
-	if err != nil {
-		return fmt.Errorf("Failed to identify user %v: %v", userid, err)
+// Signal sends a signal to the process managed by the executor
+func (l *LibcontainerExecutor) Signal(s os.Signal) error {
+	return l.userProc.Signal(s)
+}
+
+// Exec starts an additional process inside the executor
+func (l *LibcontainerExecutor) Exec(deadline time.Time, cmd string, args []string) ([]byte, int, error) {
+	combined := append([]string{cmd}, args...)
+	// Capture output
+	buf, _ := circbuf.NewBuffer(int64(dstructs.CheckBufSize))
+
+	process := &libcontainer.Process{
+		Args:   combined,
+		Env:    l.command.Env,
+		Stdout: buf,
+		Stderr: buf,
 	}
 
-	// Get the groups the user is a part of
-	gidStrings, err := u.GroupIds()
+	err := l.container.Run(process)
 	if err != nil {
-		return fmt.Errorf("Unable to lookup user's group membership: %v", err)
+		return nil, 0, err
 	}
 
-	gids := make([]uint32, len(gidStrings))
-	for _, gidString := range gidStrings {
-		u, err := strconv.Atoi(gidString)
-		if err != nil {
-			return fmt.Errorf("Unable to convert user's group to int %s: %v", gidString, err)
+	waitCh := make(chan *waitResult)
+	defer close(waitCh)
+	go l.handleExecWait(waitCh, process)
+
+	select {
+	case result := <-waitCh:
+		ps := result.ps
+		if result.err != nil {
+			if exitErr, ok := result.err.(*exec.ExitError); ok {
+				ps = exitErr.ProcessState
+			} else {
+				return nil, 0, result.err
+			}
+		}
+		var exitCode int
+		if status, ok := ps.Sys().(syscall.WaitStatus); ok {
+			exitCode = status.ExitStatus()
 		}
+		return buf.Bytes(), exitCode, nil
 
-		gids = append(gids, uint32(u))
+	case <-time.After(time.Until(deadline)):
+		process.Signal(os.Kill)
+		return nil, 0, context.DeadlineExceeded
 	}
 
-	// Convert the uid and gid
-	uid, err := strconv.ParseUint(u.Uid, 10, 32)
-	if err != nil {
-		return fmt.Errorf("Unable to convert userid to uint32: %s", err)
-	}
-	gid, err := strconv.ParseUint(u.Gid, 10, 32)
-	if err != nil {
-		return fmt.Errorf("Unable to convert groupid to uint32: %s", err)
+}
+
+type waitResult struct {
+	ps  *os.ProcessState
+	err error
+}
+
+func (l *LibcontainerExecutor) handleExecWait(ch chan *waitResult, process *libcontainer.Process) {
+	ps, err := process.Wait()
+	ch <- &waitResult{ps, err}
+}
+
+func configureCapabilities(cfg *lconfigs.Config, command *ExecCommand) {
+	// TODO: allow better control of these
+	cfg.Capabilities = &lconfigs.Capabilities{
+		Bounding:    allCaps,
+		Permitted:   allCaps,
+		Inheritable: allCaps,
+		Ambient:     allCaps,
+		Effective:   allCaps,
 	}
 
-	// Set the command to run as that user and group.
-	if e.cmd.SysProcAttr == nil {
-		e.cmd.SysProcAttr = &syscall.SysProcAttr{}
+}
+
+func configureIsolation(cfg *lconfigs.Config, command *ExecCommand) {
```

Comment:
More comments in this function would be nice to explain what each step is doing (and why these values were chosen)